docs/GDPs/gdp-0002-regions.rst

```diff
@@ -0,0 +1,275 @@
+
+.. code-block:: python
+
+    istart = I[0] if grid.west_edge else None
+    jstart = J[0] if grid.south_edge else None
+    iend = I[-1] if grid.east_edge else None
+    jend = J[-1] if grid.north_edge else None
+
+    @gtscript.stencil()
+    def stencil(uc: Field, vc: Field, cosa: Field, rsina: Field, ut: Field, ub: Field, dt5: float, dt4: float):
+        from __externals__ import istart, iend, jstart, jend
+        with computation(PARALLEL), interval(...):
+            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with horizontal(region[istart, :], region[iend, :]):
+                ub = dt5 * (ut[0, -1, 0] + ut)
+            with horizontal(region[:, jstart], region[:, jend]):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+Notice that distributed domain decomposition works with this feature - if the external has the value ``None``, then any ``region`` objects using this in its slices ignores it.
+
+This greatly reduces the complexity of the code and consolidates operations on ``ub`` - it is now immediately clear what the stencil is filling into ``ub`` everywhere.
+
+
+Usage and Impact
+----------------
+
+This is an optional feature, but will be the accepted approach to specialize computation at points in the horizontal iteration space.
+
+
+Backward Compatibility
+----------------------
+
+This GDP aims to be fully backward-compatible.
+
+
+Detailed Description
+--------------------
+
+As introduced above, we propose adding a new ``with horizontal()`` context that specializes the stencil on a region of the horizontal axes bounds using ``region`` objects, which pass information to GT4Py about the horizontal iteration space through the indexing operator, similar to numpy arrays.
+
+
+Axis Offsets
+++++++++++++
+
+Regions computation is specified using `Axis Offsets`, which are defined in GT4Py by subscripting the axes (``I``, ``J``, and ``K``).
+These can be indexed, which returns the specific indices within a stencil relative to the compute origin. For example: ``I[0]`` is the first compute point, ``I[1]`` the second, and finally ``I[-1]`` is the last point in the stencil compute domain along the ``I`` axis.
+
+Stencil computation in the horizontal axes behaves differently than in the vertical because statements execute over an index space that may extend beyond the limits defined in the stencil compute domain.
+Such ``extents`` cannot be represented by merely subscripting axes, since for example ``I[-1]`` referes to the last compute domain index along the ``I`` axis, not the point before the beginning of it.
+Axis Offsets therefore interally hold another offset, which can be set by adding or subtracting from the subscript.
+For example ``I[0] - 2`` is itself an Axis Offsets that refers to 2 points before the start of the compute domain.
+
+Axis Offsets may be assigned to variables in Python and/or used as externals in GT4Py in order to define ``region``.
```

Comment:
Do you mean that an AxisOffset object can be created and manipulated in Python or that AxisOffset indices can be Python variables? It's not 100% clear to me.

Reply:
Both. They are a proper class with `__add__` `__sub__`, and the r-variants.